### DISCUSSION TITLE
Rotation matrix to convert AirSim local coordinates to global coordinates

### DISCUSSION THREAD

**Discussion by AngelTang190** [microsoft/AirSim#4283] [Support Q&A]
### What's your question?

Is there a rotation matrix to convert local AirSim coordinates to global coordinates and vice versa?

### Include context on what you are trying to achieve

N/A

#### Context details

This question is only applicable to AirSim version only.

### Include details of what you already did to find answers
I tried to search online about what is the rotation matrix to convert between AirSim coordinates and global coordinates, but I couldn't find any, seems like it needs to compare the direction of AirSim and the Earth. For example, the North of AirSim may be different from the True North (which is the Earth's North), so the rotation matrix synchronizes both the North, but I'm not sure about this. Can anyone please help? Thank you!

**Comment by ahmed-elsaharti** [comment]
Hi @AngelTang190 ,
AFAIK, what you're trying to achieve will mainly depend on how the Unreal Engine level/environment is set up. i.e.: if your unreal environment is set up so that it resembles an actual location on earth, the angle between where (in the map/environment) north would be and the x-axis (for example) would mainly depend on how your environment is rotated within Unreal Engine.

Things are slightly simpler when it comes to GPS points though. You specify where exactly you want your drone to spawn ([X, Y and Z](https://github.com/microsoft/AirSim/blob/master/docs/settings.md#common-vehicle-setting)) within the map in [settings.json](https://github.com/microsoft/AirSim/blob/master/docs/settings.md) (or in the Unreal Editor by moving `PlayerStart`) and assign that spawn location to a real-world GPS coordinate (see [OriginGeopoint](https://github.com/microsoft/AirSim/blob/master/docs/settings.md#origingeopoint))

In general, if you are trying to line up the map's North and AirSim's North **visually**, you might need to do that from within the environment itself.
But if you don't care how it looks in the viewport, AFAIK, AirSim defaults the environment's x-axis to correspond to North. So If you track GPS data and move the drone in the x-direction, you'll see that it's moving in the north direction.

In the test below I used Droneshell to move the drone in the x-direction and it corresponded to motion in the north direction from where my OriginGeopoint/spawn-point was 
![image](https://user-images.githubusercontent.com/49626509/102606782-ebf39200-40ec-11eb-90be-d2685b84f52d.png)

**Comment by AngelTang190** [comment]
May I know how can I know the orientation of AirSim environment? Is there an API that I can use ?

**Comment by ahmed-elsaharti** [comment]
@AngelTang190 the [getMultirotorstate](https://microsoft.github.io/AirSim/apis/#getmultirotorstate) and/or simGetGroundTruthKinematics APIs would provide the orientation in quaternions

**Comment by AngelTang190** [comment]
Can you please give an example? As in when an orientation has a x value of 0 initially, what does it mean?

**Comment by ahmed-elsaharti** [comment]
@AngelTang190 a yaw orientation of 0 means that the front of the vehicle is aligned with the World's/environment's x axis. Note that the multirotor state produces orientations in [quaternions](https://en.wikipedia.org/wiki/Quaternion) (for conversion see [this](https://quaternions.online/))

Here's an example that should clarify everything:

Lets say we have a blank unreal engine environment/level:
![image](https://user-images.githubusercontent.com/49626509/102649646-29790f00-412f-11eb-9ccb-76ae0f145b39.png)

This environment would have a set of x,y and z axes that start from a point (the world origin). I'll represent the environment's top view in my diagrams with this grid:
![image](https://user-images.githubusercontent.com/49626509/102652427-752db780-4133-11eb-85ed-41d310717a4f.png)

Now, when u place the AirSim plugin into the Unreal Engine project, the x-axis is considered north (keep in mind that there is no map or anything here yet, its just an empty environment, and north is just relative to the environment's coordinate system):
![image](https://user-images.githubusercontent.com/49626509/102652854-b1f9ae80-4133-11eb-995b-f46d196aa125.png)
And here is our setup with a drone placed at (3,4,0) (Note that if you set an OriginGeopoint in your settings.json like: 
```
"OriginGeopoint": {
    "Latitude": 47.641468,
    "Longitude": -122.140165,
    "Altitude": 122
  }
``` 
your drone spawn point will be set to the coordinates you enter):
![image](https://user-images.githubusercontent.com/49626509/102653795-1f5a0f00-4135-11eb-8219-aedbde830f60.png)

Again, this is all without having anything related to 'real earth' in our environment.

Now, since the angle between the forward direction of the drone (the blue arrow below) and the world's x axis is 0. Your yaw value would be 0 and your orientation in [quaternions](https://quaternions.online/) would be [ 0, 0, 0, 1 ].
![image](https://user-images.githubusercontent.com/49626509/102654555-5250d280-4136-11eb-82af-87e0972e826a.png)

If the drone is spawned at an angle (pitch, yaw or roll):
![image](https://user-images.githubusercontent.com/49626509/102654913-02bed680-4137-11eb-96de-e6c881f6360c.png)
the angles between the drone and the world's coordinate system will be reported in the state. (-45 degree yaw in the diagram).

Now this behavior should not change regardless of what we place in the environment. For example, if I place assets in the UE environment that represents a particular area on earth but these assets/map-items are not alligned with the world coordinate system:
![image](https://user-images.githubusercontent.com/49626509/102658043-6c8daf00-413c-11eb-9546-758d713deecc.png)
There would be a difference between the world/game's coordinate system's North and the 'maps' "true north". If you know that difference (assuming you are the one building the level/environment so you would know what the offset is) you can always add it to your calculations. 
OR you could just build your environment so that it's lined up correctly:
![image](https://user-images.githubusercontent.com/49626509/102658244-d0b07300-413c-11eb-8f09-50e60a2ddfa7.png)

Hope this clarifies things. I might be mistaken in how exactly the aligning works within UE but this is how my understanding of it is.

**Comment by AngelTang190** [comment]
Alright, thank you very much! So let's say if the yaw is 0, then the rotation matrix will just be the identity matrix correct? 
(Since there is no rotation and I want the x axis to point to the North?)

**Comment by ahmed-elsaharti** [comment]
@AngelTang190 if you're refering to a 2D rotation matrix about the Z axis then yes. But if you're talking about an overall 3D rotation matrix then pitch and roll would have to be taken into consideration

**Comment by AngelTang190** [comment]
For 3D rotation matrix, should pitch also be aligned with the North? Then which direction should roll be aligned to?

**Comment by ahmed-elsaharti** [comment]
Pitch would be measured based on the horizon or the vehicle being upright. Same thing with roll. They would both be relative to the drone's upright orientation

**Comment by AngelTang190** [comment]
May I know what does upright mean here?

**Comment by petergerten** [comment]
I am currently using AirSim together with the Cesium for Unreal plugin. Cesium puts a real world map into the Unreal project. I used the same coordinates for the georeference origin in AirSim and Cesium. This works, the drone shows at the correct position and GPS data (PX4 SITL) is correct. But when I fly into any directions - they don't line up. I am a bit lost where I would configure that. Any hints?

**Comment by ahmed-elsaharti** [comment]
@petergerten by 'they don't line up' do you mean there's an angle offset? or is it a distance offset?

**Comment by petergerten** [comment]
@ahmed-elsaharti I need to evaluate it more closely. But it seems distance is correct but when I fly south in the airsim real-world environment - my GPS coordinates in AirSim/PX4 go east.

**Comment by ahmed-elsaharti** [comment]
@petergerten hmmm, that probably is just a difference in rotation between airsims and cesiums coordinate systems (or atleast i hope its just that). Try rotating the playerstart object within the environment and see if that fixes things.

I haven't tested cesium out yet... But this is very interesting

**Comment by petergerten** [comment]
@ahmed-elsaharti I tried to change the rotation of player start - but it does not seem to have an impact on the world rotation.

**Comment by petergerten** [comment]
ok, after rotating the Cesium World Terrain 90 degrees - everything lines up :)

**Comment by salvaba94** [comment]
@ahmed-elsaharti @petergerten  I am trying to do a similar thing but linking the Cesium Georeference longitude and latitude to the AirSim OriginGeopoint settings so that the point in the planet can be configured. Do you know how this can be accomplished?

**Comment by leanedonze** [comment]
I know this issue is a bit old, but I am encountering a similar problem.
I rotated the Cesium World Terrain by 90 degrees and AirSim and Cesium coordinate systems are now aligned.
However, because of the rotation, any object placed with a Cesium Globe Anchor is moved and not at the correct location anymore. I would expect the Cesium Globe Anchor object to take care of this for me, but it doesn't.
Does anyone have an idea for this?
